### PR TITLE
Ensemble of neural posteriors

### PR DESCRIPTION
**Pull request by florpi**
Added flexibility in config files to produce an ensemble of neural posteriors. Now the config file takes in a list of posterior configurations and produces an ensemble posterior model, where each of the individual posteriors are weighted by their corresponding validation losses.

### CODE REVIEW

**Comment by florpi** [comment]
It looks like ensembling 4 models (2 mafs and 2 mixture density nets) we find more stable results when re-initializing the model. Here you can see a different validation example in each row, whereas the first column are models with one single maf, and the second column is the result of ensembling 4 models. v1 and v2 are the same model trained on the same data, but with different random initializations.

|alt |alt|
|-|-|
|alt | alt

File: ili/inference/runner_sbi.py [hunk 1]
```diff
@@ -99,48 +109,76 @@ def load_object(cls, config: Dict) -> Any:
...
-        posterior = self.model.build_posterior()
-        # save posterior
+        posteriors, val_loss = [], []
+        for n, posterior in enumerate(self.neural_posteriors):
+            logging.info(
+                f"Training model {n+1} out of {len(self.neural_posteriors)} ensemble models"
+            )
+            model = self.inference_class(
+                prior=self.prior,
+                density_estimator=posterior,
```
Comment (maho3) [inline ili/inference/runner_sbi.py:167]: Only works for SNPE_C

File: ili/inference/runner_sbi.py [hunk 2]
```diff
@@ -99,48 +109,76 @@ def load_object(cls, config: Dict) -> Any:
...
+
+        Args:
+            inference_module (str): module from which to import class
+            inference_class (str): class name
+
+        Returns:
+            NeuralInference: neural inference class
+        """
+        module = importlib.import_module(inference_module)
+        return getattr(module, inference_class)
```
Comment (maho3) [inline ili/inference/runner_sbi.py:127]: In this, we can't specify additional arguments of the inference class. e.g. logging_level

**Comment by maho3** [comment]
Really great code! Only big problem is we've drifted away from including functionality for all sbi inference classes. Only SNPE_C works under this. Here's an example sample_inference_config.yaml in which SNRE_A breaks:

```
loader:
  module: 'ili.inference.loaders'
  class: 'StaticNumpyLoader'
  args:
    in_dir: 'toy'
    x_file: 'x.npy'
    theta_file: 'theta.npy'

prior:
  module: 'sbi.utils'
  class: 'BoxUniform'
  args:
    low: [0,0,0]
    high: [1,1,1]

model:
  module: 'sbi.inference'
  class: 'SNRE_A'
  neural_posteriors:
    - classifier: 'linear'
    - classifier: 'mlp'
    
train_args:
  training_batch_size: 32
  learning_rate: 0.001

device: 'cpu'
output_path: 'toy'
```

I think if instead of training the posterior_nn within the same inference class (e.g. SNPE_C), we could initialize a list of inference class objects (e.g. [SNPE_C_1, SNPE_C_2]) and train separately.

I'm also not sure if e.g. SNRE._build_posterior works with NeuralPosteriorEnsemble, but its something to check before we push to main.

**Comment by modichirag** [comment]
This looks great and the example figure above is quite promising. I did not know about NeuralPosteriorEnsemble wrapper, that makes life very easy.

Some thoughts- 

- As Matt said, it will be good to check if SNRE is expected to work with ensemble setup.

- Regarding SNPE_A/C1/C2, if I understand things correctly, they differ only if we are doing multiple rounds of inference i.e. once we get a first posterior, we run more simulations in the posterior regime. That means active sampling and if we are setting up the code with that in mind, we will need to do broader changes.

**Comment by modichirag** [comment]
The current setup is great to begin with and start playing around. So the following thoughts are probably not relevant for this PR, but I think it would be good to consider them at some point soon in the future.
I think we will need to expand the current setup in a few ways (I am trying to implement a few of these on my pipeline, happy to work on implementing them here as well)- 

1. currently training and ensembling of posteriors is done together. We probably want to separate it. It will be good to save every posterior individually and then combining them together as ensembles in a different function. This allows us to expand and cull ensembles later, experiment with different weighings and such. 

2. Most importantly, this^ allows us to run parallel jobs to train multiple posteriors instead of training them sequentially. 

3.  Each configuration currently needs to be written explicitly in the config file. This might make actual hyperparameter search intractable, where we will want to try permutations of different parameters.
Two ways around this are- generating a new single model parameter file from a script for each permutation and train a posterior for each, or to accept distributions over parameters in config file and sample from this in the runner to train multiple models

4. Following from above, we will likely end up training more models for hyperparameter search than we actually use in the final ensemble, so separating training and ensembling helps with this.

**Comment by modichirag** [comment]
Another point that a colleague pointed out to me last week was that by default SBI uses different validation set for every training, and so the validation losses are estimated on different simulations. It will probably not affect our results significantly, but something to be mindful of and we probably want to address at some point.

**Review by modichirag — APPROVED** [review]
